### PR TITLE
fix: handle Anthropic-style tool schemas in LangchainProfilerHandler

### LINKED ISSUE #432: [BUG]: LangchainProfilerHandler doesn't handle ChatAnthropic tools
### Version

latest

### Which installation method(s) does this occur on?

_No response_

### Describe the bug.

`LangchainProfilerHandler` implementation constraints with OpenAI Tool Schema.

https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/540408cdc192f2bdcccbaedbded7756086e1e1ff/src/aiq/profiler/callbacks/langchain_callback_handler.py#L51

 If I use toolkit with custom llm provider like `langchain_anthropic.ChatAnthropic`, a schema mismatch occurs.

```output
langchain_core.callbacks.manager - WARNING - Error in LangchainProfilerHandler.on_chat_model_start callback: 2 validation errors for ToolSchema
```

### Minimum reproducible example

```shell

```

### Relevant log output

<details><summary>Click here to see error details</summary><pre>

 [Paste the error here, it will be hidden by default]

</pre></details>

### Other/Misc.

_No response_

### Code of Conduct

- [x] I agree to follow the AIQ toolkit Code of Conduct
- [x] I have searched the [open bugs](https://github.com/NVIDIA/NeMo-Agent-Toolkit/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

### PR DESCRIPTION
**Pull request by cluster2600**
## Description

`_extract_tools_schema` only accepted OpenAI-format tool definitions (nested `type`/`function` keys).
When using `ChatAnthropic`, tools are sent with top-level `name`/`description`/`input_schema` keys,
which caused a silent parse failure — the tool schema was quietly dropped and never recorded in the
trace metadata.

This adds a fallback path that converts the Anthropic layout into the canonical `ToolSchema` model,
so tool metadata is preserved regardless of which LLM provider is being used.

Closes #432

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://nvidia.github.io/NeMo-Agent-Toolkit/).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

## Test plan

- Added four unit tests to `test_langchain_callback_handler.py`:
  - `test_extract_tools_schema_openai_format` — verifies the existing behaviour with OpenAI-style tools
  - `test_extract_tools_schema_anthropic_format` — exercises the new Anthropic fallback path
  - `test_extract_tools_schema_mixed_formats` — confirms both formats work together in a singel request
  - `test_extract_tools_schema_empty_and_none` — edge cases (empty list, `None`)
- Manually verified that an Anthropic tool dict like `{"name": "search", "description": "...", "input_schema": {...}}` is correctly converted into a `ToolSchema(type="function", function=ToolDetails(...))`.

## Changes

- `callback_handler.py`: added imports for `ToolDetails` / `ToolParameters`; wrapped the existing `ToolSchema(**tool)` call in a try/except that falls back to manual conversion from Anthropic format.  If both conversions fail the tool is logged at DEBUG level and skipped, matching the previous behaviour.

## Summary by CodeRabbit

* **New Features**
  * Improved tool schema compatibility to support multiple AI tool formats (OpenAI and Anthropic styles) with more robust fallback parsing for broader integration.

* **Bug Fixes**
  * Hardened error handling: fallback parsing failures now log exceptions to surface unexpected tool formats.

* **Tests**
  * Added comprehensive tests for OpenAI, Anthropic, mixed formats, additional-properties, unparseable inputs, and other edge cases.

### CODE REVIEW

**Comment by willkill07** [comment]
@cluster2600 please ensure that CI / pre-commit checks pass locally. When I triggered CI there's code formatting that's failed.

**Comment by cluster2600** [comment]
Hi @willkill07, thanks for flagging the CI failure — I've pushed a fix (d4c38ea) that addresses the `yapf` formatting. The inline dict literals in the test file have been expanded to multi-line format as per the project's style config.

Verified locally:
- `yapf --style ./pyproject.toml --diff` returns clean on both changed files
- `ruff check` passes 
- All 7 tests in `test_langchain_callback_handler.py` pass on Python 3.12

Let me know if there's anythng else needed!

File: packages/nvidia_nat_langchain/tests/test_langchain_callback_handler.py [hunk 1]
```diff
@@ -87,3 +88,158 @@ async def test_langchain_handler(reactive_stream: Subject):
...
+    result = _extract_tools_schema(invocation_params)
+    assert len(result) == 1
+    assert result[0].function.name == "good_tool"
+
+
+def test_extract_tools_schema_empty_and_none():
+    """Test edge cases: empty tools list and None invocation_params."""
+    assert _extract_tools_schema({}) == []
+    assert _extract_tools_schema({"tools": []}) == []
+    assert _extract_tools_schema(None) == []
```
[the remark this answers is not in the document]
Comment (cluster2600) [reply packages/nvidia_nat_langchain/tests/test_langchain_callback_handler.py:280]: Good catch — I'd overlooked that a truthy non-mapping value (e.g. `[{"type": "string"}]`) slips past the `or {}` guard and raises `AttributeError` on the subsequent `.get()` call, which wasn't being caught.

Added the test for this case and also included `AttributeError` in the except clause alongside `KeyError` and `TypeError` so that the fallback properly logs and skips the malfomed tool rather than letting the exception escape.

Addressed in 0d2e1262.

**Comment by cluster2600** [comment]
@willkill07 Cheers — pushed another small fix (a8334b9). Turns out there was one remaining `yapf` violation I'd missed: an inline dict literal inside a list in `test_extract_tools_schema_skips_non_mapping_input_schema` that needed expanding to multi-line format (the `split_all_top_level_comma_separated_values` rule).

Verified locally that both changed files now pass `yapf --style ./pyproject.toml --diff` cleanly, and `ruff check` is green as well. Should be all sorted — happy to adress anything else if needed.

File: packages/nvidia_nat_langchain/src/nat/plugins/langchain/callback_handler.py [hunk 2]
```diff
@@ -52,11 +54,27 @@ def _extract_tools_schema(invocation_params: dict) -> list:
...
+                                    required=input_schema.get("required", []),
+                                ),
+                            ),
+                        ))
+                except Exception:
+                    logger.debug(
+                        "Failed to parse tool schema from invocation params: %s. \n This "
+                        "can occur when the LLM server has native tools and can be ignored if "
+                        "using the responses API.",
+                        tool)
```
[the remark this answers is not in the document]
Comment (cluster2600) [reply packages/nvidia_nat_langchain/src/nat/plugins/langchain/callback_handler.py:78]: Addressed in 292acf0 — tightened the inner except to `(KeyError, TypeError, AttributeError)` and switched to `logger.exception()` for full stack traces. Also included `AttributeError` since a non-mapping `input_schema` (e.g. a list) would blow up on the subsequant `.get()` call, which I caught in a later commit.

File: packages/nvidia_nat_langchain/src/nat/plugins/langchain/callback_handler.py [hunk 3]
```diff
@@ -52,11 +54,27 @@ def _extract_tools_schema(invocation_params: dict) -> list:
...
+                    tools_schema.append(
+                        ToolSchema(
+                            type="function",
+                            function=ToolDetails(
+                                name=tool["name"],
+                                description=tool.get("description", ""),
+                                parameters=ToolParameters(
+                                    properties=input_schema.get("properties", {}),
+                                    required=input_schema.get("required", []),
+                                ),
```
[the remark this answers is not in the document]
Comment (cluster2600) [reply packages/nvidia_nat_langchain/src/nat/plugins/langchain/callback_handler.py]: Good spot — this was indeed missing. Addressed in 292acf0: the `ToolParameters` constructor now explicitly passes `additionalProperties=input_schema.get("additionalProperties", False)` so the original schema contract is faithfully presevred in trace metadata. Also added `test_extract_tools_schema_anthropic_additional_properties` to lock in the behaviour.

**Comment by cluster2600** [comment]
Re the nitpick from the latest review round (adding `caplog: pytest.LogCaptureFixture` type hint) — fair point, though none of the other test functions in this file carry parameter annotations and `pytest` isn't currently imported as a module (only its fixtures are injected). I'd rather keep the style consistant with the rest of the test suite for now and address type hints across all test files in a separate cleanup pass.

**Comment by willkill07** [comment]
@cluster2600 you are welcome to ignore any code rabbit feedback as we know it's not perfect. Re-triggering CI now

**Comment by willkill07** [comment]
/ok to test [08a42a0](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/1770/commits/08a42a06633c931dc836fb4dd55a1a42953ffceb)

**Review by willkill07 — APPROVED** [review]

**Comment by cluster2600** [comment]
The CI Pipeline / Check failure is caused by pre-existing dead links in the `dynamo_integration` example documentation, not by the changes in this PR:

```
ERROR: 1 dead link found in examples/dynamo_integration/README.md !
ERROR: 1 dead link found in examples/dynamo_integration/react_benchmark_agent/README.md !
ERROR: 1 dead link found in external/dynamo/README.md !
```

This affects other open PRs as well (e.g. #1775). The linter and documentation checks both pass — only the dead link checker is triggering the faliure.

**Comment by willkill07** [comment]
@cluster2600 I'm aware. I had a PR merged last night that addressed that and just merged `develop` into your branch.

**Comment by willkill07** [comment]
/ok to test [f0106d3](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/1770/commits/f0106d3d569bf89366c77d507e839ac0e6b905ed)

**Comment by willkill07** [comment]
@cluster2600 please do not update the branch as it requires me to retrigger CI

**Review by willkill07 — APPROVED** [review]

**Comment by willkill07** [comment]
/ok to test [d32a2f9](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/1770/commits/d32a2f91a0923890223c1c1fcdcf61c4e1d8042c)
